corehq/util/models.py

```diff
@@ -241,3 +240,65 @@ def to_python(self, value):
     def pre_init(self, value, obj):
         value = super(NullJsonField, self).pre_init(value, obj)
         return self.get_default() if value is None else value
+
+
+class ForeignValue:
+    """Property descriptor for Django foreign key refs with a primitive value
+
+    This is useful for cases where the object referenced by the foreign
+    key holds a single primitive value named `value`. It eliminates
+    boilerplate indirection imposed by the foreign key, allowing the
+    value to be referenced as a simple attribute.
+    """
+
+    def __init__(self, foreign_key: models.ForeignKey, truncate=False):
+        self.fk = foreign_key
+        self.truncate = truncate
+
+    def __set_name__(self, owner, name):
+        other_names = getattr(owner, "_ForeignValue_names", [])
+        owner._ForeignValue_names = other_names + [name]
+
+    def __get__(self, obj, objtype=None):
+        if obj is None:
+            return self
+        fobj = getattr(obj, self.fk.name)
+        return fobj.value if fobj is not None else None
+
+    def __set__(self, obj, value):
+        if value is None:
+            if getattr(obj, self.fk.name) is not None:
+                setattr(obj, self.fk.name, None)
+            return
+        model = self.fk.related_model
+        if self.truncate:
+            maxlen = model._meta.get_field("value").max_length
+            value = value[:maxlen]
+        fobj, _ = model.objects.get_or_create(value=value)
+        setattr(obj, self.fk.name, fobj)
+
+    @staticmethod
+    def get_names(cls):
+        """Get a list of ForeignValue attribute names of the given class
+
+        Raises `AttributeError` if the class has no `ForeignValue` attributes.
+        """
+        return cls._ForeignValue_names
+
+
+def foreign_value_init(cls):
```

Comment:
This definitely seems very complex to me (the whole `ForeignValue` setup). The caching in particular is very hard to reason about since we're mixing an LRU cache with a cached property (does it even live beyond the individual objects lifespan?).

I feel like a simple utility function or two would suffice and then adding methods to the model classes that use the utility functions.

Reply:
You only need this decorator if you want to pass foreign values as `__init__` arguments. I do understand the argument for using a base class instead of a decorator, although that limits the type hierarchy in other ways that doesn't seem necessary for this functionality. Implementing it as a mixin seems less than ideal because the ordering of the mixin with other base classes is very important and easy to get wrong.

Yes, a custom foreign key field implementation could probably avoid this, although that seemed more complicated.

I'm inclined to leave it for now and possibly refactor it in a follow-up PR, but probably only if it gets used more widely.

Reply:
> This definitely seems very complex to me (the whole ForeignValue setup).

To me it doesn't seem much more complicated than what's happening inside of Django's `ForeignKey` field. I tried to encapsulate the logic within a property that hopefully has clear semantics for an admittedly niche use case. Ultimately it's an optimization for a few large values associated with in a high-volume table. I think all important functionality is tested, so I think it's pretty solid.

> I feel like a simple utility function or two would suffice and then adding methods to the model classes that use the utility functions.

I didn't see another way to accomplish the same thing as succinctly without duplicating a lot of the logic on every class that uses `ForeignValue`. Can you elaborate on the simple utility functions that you're thinking of?

> mixing an LRU cache with a cached property (does it even live beyond the individual objects lifespan?)

Yes. The cache is stored on the propery instance, which is associated with the class object. It uses `cached_property` so the LRU cache is not re-created each time the property is accessed. The cache lives for the duration of the interpreter lifetime from when the property is first accessed on an instance object. That part is more complicated than I wanted, but I tried to cover it with tests to ensure it works as expected.

Reply:
Again, I'll defer to your judgment. I appreciate that this lets me use these models in a more efficient fashion, but it still feels pretty cute...

Reply:
> > This definitely seems very complex to me (the whole ForeignValue setup).
> 
> To me it doesn't seem much more complicated than what's happening inside of Django's `ForeignKey` field. I tried to encapsulate the logic within a property that hopefully has clear semantics for an admittedly niche use case. Ultimately it's an optimization for a few large values associated with in a high-volume table. I think all important functionality is tested, so I think it's pretty solid.

For sure it's not more complex than that but it's a lot of complexity for a very niche use case and it does make it hard to reason about what's going on. The Django FK models are well known and understood and match what's happening at the DB layer, this is not. 
 
> > I feel like a simple utility function or two would suffice and then adding methods to the model classes that use the utility functions.
> 
> I didn't see another way to accomplish the same thing as succinctly without duplicating a lot of the logic on every class that uses `ForeignValue`. Can you elaborate on the simple utility functions that you're thinking of?

Sure there would be some duplication but since this is essentially just an abstraction on a lookup you could do something like this:
```python
def get_user_agent(self):
    return get_value_from_foreign_key(UserAgent, self.user_agent_fk_id)

def set_user_agent(self, value):
   return self.user_agent_fk = get_or_create_value(UserAgent, value)
```

> > mixing an LRU cache with a cached property (does it even live beyond the individual objects lifespan?)
> 
> Yes. The cache is stored on the propery instance, which is associated with the class object. It uses `cached_property` so the LRU cache is not re-created each time the property is accessed. The cache lives for the duration of the interpreter lifetime from when the property is first accessed on an instance object. That part is more complicated than I wanted, but I tried to cover it with tests to ensure it works as expected.

Ah, I had it in my head that this was on the model instance but it's on they `ForeignValue` instance which has a different lifecycle to the model. Still, I think that just illustrates how hard this can be to reason about.

Considering that ALL usages of this make use of the cache I would recommend simplifying the code to make caching mandatory instead of optional.

Reply:
Upon revisiting this after the initial rollout drama, I'm going to keep the `cache_size` parameter since it was useful to be able to disable the cache temporarily during a database migration.

While I concede that there is a fair amount going on in `ForeignValue`, I still think it is a useful abstraction neatly encapsulated within a property descriptor. I think it is significant performance optimization for auditcare, saving up to two database round-trips per request.

It may make sense to limit its availability scope by moving it into the auditcare app if/when `UserAccessLog` gets merged into auditcare.